hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneSecurityUtil.java

```diff
@@ -141,4 +146,39 @@ public static List<X509Certificate> convertToX509(
     }
     return x509Certificates;
   }
+
+  public static void updateKerberosInfo(Class clz, String serverPrincipal) {
+    // new KerberosInfo
+    Annotation newValue = new KerberosInfo() {
+      @Override
+      public Class<? extends Annotation> annotationType() {
+        return KerberosInfo.class;
+      }
+      @Override
+      public String serverPrincipal() {
+        return serverPrincipal;
+      }
+      @Override
+      public String clientPrincipal() {
+        return null;
+      }
+    };
+
+    try {
+      Method method = clz.getClass().getDeclaredMethod("annotationData", null);
+      method.setAccessible(true);
+
+      Object annotationData = method.invoke(clz);
+      Field annotations = annotationData.getClass().getDeclaredField("annotations");
+      annotations.setAccessible(true);
+
+      Map<Class<? extends Annotation>, Annotation> map =
+          (Map<Class<? extends Annotation>, Annotation>) annotations.get(annotationData);
+      map.put(KerberosInfo.class, newValue);
```

Comment:
I'd like to avoid setting the annotation via reflection.  Can we use separate `ReconfigureProtocolPB` interfaces for different services, similar to how `SecretKeyProtocol...PB` is defined?

https://github.com/apache/ozone/blob/5f6306dd1dde3c9c982ed933aa1fa8f4fa7a9301/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java#L172-L177

https://github.com/apache/ozone/blob/5f6306dd1dde3c9c982ed933aa1fa8f4fa7a9301/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocolPB/SecretKeyProtocolDatanodePB.java#L33-L38

Reply:
@adoroszlai , thanks for the review. 
Use different interface is another approach to solve this problem. I considered it as one of the choice at the beginning. 
Any specific reason that reflection is not favored? As far as I can see, it's used in many places of Ozone now.

Reply:
@ChenSammi we can create follow-up task to replace reflection with some better way to use specific principal